Better handling of malformed config files

When trying to run the agent I started it once with a malformed config file:
```yaml
name: retrofit
packages:
path: retrofit2
```

This led to a scary-looking and uninformative crash:
```console
$ cd /home/divide/projects/tmp/retrofit ; /usr/bin/env /usr/lib/jvm/java-17-openjdk-amd64/bin/java -agentlib:jdwp=transport=dt_socket,server=n,suspend=y,address=localhost:40273 @/tmp/cp_6nllvygw5957cfgoemsb42vux.argfile com.example.retrofit.SimpleService                                                                                                                       
Exception in thread "main" java.lang.reflect.InvocationTargetException                                                                   
       at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                                
       at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)                              
       at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                      
       at java.base/java.lang.reflect.Method.invoke(Method.java:569)                                                                    
       at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:491)                       
       at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:503)                      
Caused by: java.lang.NullPointerException: Cannot read the array length because "com.appland.appmap.config.AppMapConfig.singleton.packages" is null                                                                                                                                
       at com.appland.appmap.config.AppMapConfig.load(AppMapConfig.java:141)                                                            
       at com.appland.appmap.config.AppMapConfig.initialize(AppMapConfig.java:308)                                                      
       at com.appland.appmap.Agent.premain(Agent.java:57)                                                                               
       ... 6 more                                                                                                                       
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message Outstanding error when calling method in invokeJavaAgentMainMethod at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 619                                                      
*** java.lang.instrument ASSERTION FAILED ***: "success" with message invokeJavaAgentMainMethod failed at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 459                                                                                                  
*** java.lang.instrument ASSERTION FAILED ***: "result" with message agent load/premain call failed at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 422                                                                                                     
FATAL ERROR in native method: processing of -javaagent failed, processJavaStart failed                                                   
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)                                                          
V  [libjvm.so+0x8e9a90]  jni_FatalError+0xb0                                                                                             
V  [libjvm.so+0xa6f0cd]  JvmtiExport::post_vm_initialized()+0x26d                                                                        
V  [libjvm.so+0xf09eca]  Threads::create_vm(JavaVMInitArgs*, bool*)+0x90a                                                                
V  [libjvm.so+0x9036d5]  JNI_CreateJavaVM+0x55                                                                                           
C  [libjli.so+0x3aa3]  JavaMain+0x93                                                                                                     
C  [libjli.so+0x80bd]  ThreadJavaMain+0xd                                                                                                
                                                                                                                                        
Przerwane  
```

--- a/agent/src/main/java/com/appland/appmap/config/AppMapConfig.java
+++ b/agent/src/main/java/com/appland/appmap/config/AppMapConfig.java
@@ -131,9 +131,13 @@ public class AppMapConfig {
     ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
     try {
       singleton = mapper.readValue(inputStream, AppMapConfig.class);
+      if (singleton.packages == null) {
+        logger.error("AppMap: missing value for the 'packages' entry in appmap.yml");
+        return null;
+      }
     } catch (IOException e) {
       logger.error("AppMap: encountered syntax error in appmap.yml {}", e.getMessage());
-      System.exit(1);
+      return null;
     }
     singleton.configFile = configFile;
     logger.debug("config: {}", singleton);

--- a/agent/src/test/java/com/appland/appmap/config/AppMapConfigTest.java
+++ b/agent/src/test/java/com/appland/appmap/config/AppMapConfigTest.java
@@ -100,6 +100,24 @@ public class AppMapConfigTest {
         AppMapConfig.load(configFile, true);
         assertEquals("/appmap", AppMapConfig.get().getAppmapDir());
     }
+
+    @Test
+    public void loadPackagesKeyWithMissingValue() throws Exception {
+        Path configFile = tmpdir.resolve("appmap.yml");
+        final String contents = "name: test\npackages:\npath: xyz";
+        Files.write(configFile, contents.getBytes());
+        String actualErr = tapSystemErr(() -> AppMapConfig.load(configFile, false));
+        assertTrue(actualErr.contains("AppMap: missing value for the 'packages'"));
+    }
+
+    @Test
+    public void loadPackagesKeyWithScalarValue() throws Exception {
+        Path configFile = Files.createTempFile("appmap", ".yml");
+        final String contents = "name:q test\npackages: abc\n";
+        Files.write(configFile, contents.getBytes());
+        String actualErr = tapSystemErr(() -> AppMapConfig.load(configFile, false));
+        assertTrue(actualErr.contains("AppMap: encountered syntax error in appmap.yml"));
+    }
 }
 
 

--- a/agent/test/intellij/intellij.bats
+++ b/agent/test/intellij/intellij.bats
@@ -31,8 +31,7 @@ setup() {
 }
 
 @test 'it works' {
-  # TODO: remove this ideVersion pin
-  run ./gradlew -PideVersion=IC-2023.3.3 :plugin-core:test --tests 'AppMapConfigFileTest.readConfigWithPath'
+  run ./gradlew :plugin-core:test --tests 'AppMapConfigFileTest.readConfigWithPath'
   assert_success
 
   output="$(< tmp/appmap/junit/appland_config_AppMapConfigFileTest_readConfigWithPath.appmap.json)"

